internal/observability/events/context_test.go

```diff
@@ -0,0 +1,855 @@
+			name:    "no-header-or-details-in-payload",
+			noFlush: true,
+			ctx:     testCtx,
+			observationPayload: []observationPayload{
+				{},
+			},
+			wantErrMatch:    errors.T(errors.InvalidParameter),
+			wantErrContains: "you must specify either header or details options",
+		},
+		{
+			name:               "no-ctx-eventer-and-syseventer-not-initialized",
+			ctx:                context.Background(),
+			observationPayload: testPayloads,
+			wantErrMatch:       errors.T(errors.InvalidParameter),
+			wantErrContains:    "missing both context and system eventer",
+		},
+		{
+			name:    "use-syseventer",
+			noFlush: true,
+			ctx:     context.Background(),
+			observationPayload: []observationPayload{
+				{
+					header: map[string]interface{}{
+						"name": "bar",
+					},
+				},
+			},
+			header: map[string]interface{}{
+				"name": "bar",
+			},
+			observationSinkFileName: c.AllEvents.Name(),
+			setup: func() error {
+				return event.InitSysEventer(hclog.Default(), c.EventerConfig)
+			},
+			cleanup: func() { event.TestResetSystEventer(t) },
+		},
+		{
+			name:                    "simple",
+			ctx:                     testCtx,
+			observationPayload:      testPayloads,
+			header:                  testWantHeader,
+			details:                 testWantDetails,
+			observationSinkFileName: c.AllEvents.Name(),
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			assert, require := assert.New(t), require.New(t)
+			if tt.setup != nil {
+				require.NoError(tt.setup())
+			}
+			if tt.cleanup != nil {
+				defer tt.cleanup()
+			}
+			op := tt.name
+			if tt.noOperation {
+				op = ""
+			}
+			require.Greater(len(tt.observationPayload), 0)
+			for _, p := range tt.observationPayload {
+				err := event.WriteObservation(tt.ctx, event.Op(op), event.WithHeader(p.header), event.WithDetails(p.details))
+				if tt.wantErrMatch != nil {
+					require.Errorf(err, "wanted error %q", tt.wantErrMatch)
+					if tt.wantErrContains != "" {
+						assert.Contains(err.Error(), tt.wantErrContains)
+					}
+					return
+				}
+				require.NoError(err)
+			}
+			if !tt.noFlush {
+				require.NoError(event.WriteObservation(tt.ctx, event.Op(tt.name), event.WithFlush()))
+			}
+
+			if tt.observationSinkFileName != "" {
+				defer func() { _ = os.WriteFile(tt.observationSinkFileName, nil, 0o666) }()
+				b, err := ioutil.ReadFile(tt.observationSinkFileName)
+				assert.NoError(err)
+
+				gotObservation := &eventJson{}
+				err = json.Unmarshal(b, gotObservation)
+				require.NoErrorf(err, "json: %s", string(b))
+
+				actualJson, err := json.Marshal(gotObservation)
+				require.NoError(err)
+				wantJson := testObservationJsonFromCtx(t, tt.ctx, event.Op(tt.name), gotObservation, tt.header, tt.details)
+
+				assert.JSONEq(string(wantJson), string(actualJson))
+			}
+		})
+	}
+}
+
+func testObservationJsonFromCtx(t *testing.T, ctx context.Context, caller event.Op, got *eventJson, hdr, details map[string]interface{}) []byte {
```

Comment:
Just curious what `testObservationJsonFromCtx` is doing to `got` for the assertion on line 438? Are we testing that the observation going over a wire or something does not ultimately alter it?

Reply:
the broker does alter the event, so we're attempting to differentiate the meaningful from the trivial in those changes and that the event does get written to a destination sink.